### ISSUE TITLE
Going to '/admin' while using storecodes in url and a different adminhtml url will throw exception

### ISSUE THREAD

**Issue by koenner01** [magento/magento2#11140]
If you are using an adminhtml url that is different from 'admin' and set the 'Add Store Code to Urls' to 'Yes', going to '/admin' will throw an exception

### Preconditions
1. PHP7.0 
2. M2.1.9 CE (we also tested 2.1.7 CE and 2.2.0 CE)

### Steps to reproduce
1. use a different url than 'admin' for the adminhtml
2. set 'Stores' > 'Configuration' > 'General' > 'Web' > 'Add Store Code to Urls' to 'Yes'
3. go to '/admin' url

### Expected result
1. Page should go to the no-route or at least show 404

### Actual result
1. Page tries to go to the homepage with the current store set to 'admin'

We did some debugging already and found that in \Magento\Store\App\Request\PathInfoProcessor the current store is being set to 'admin'.

```
public function process(\Magento\Framework\App\RequestInterface $request, $pathInfo)
{
    $pathParts = explode('/', ltrim($pathInfo, '/'), 2);
    $storeCode = $pathParts[0];
    try {
        /** @var \Magento\Store\Api\Data\StoreInterface $store */
        $store = $this->storeManager->getStore($storeCode);
    } catch (NoSuchEntityException $e) {
        return $pathInfo;
    }
    if ($store->isUseStoreInUrl()) {
        if (!$request->isDirectAccessFrontendName($storeCode)) {
            $this->storeManager->setCurrentStore($storeCode);
            $pathInfo = '/' . (isset($pathParts[1]) ? $pathParts[1] : '');
            return $pathInfo;
        } elseif (!empty($storeCode)) {
            $request->setActionName('noroute');
            return $pathInfo;
        }
        return $pathInfo;
    }
    return $pathInfo;
}
```
'admin' will be treated as a valid storecode because the store table in the database contains this value for the store with id 0

**Comment by koenner01** [comment]
This was already reported in https://github.com/magento/magento2/issues/9947 but closed because of 'inability to reproduce'

**Comment by magento-engcom-team** [comment]
@koenner01, thank you for your report.
We've created internal ticket(s) MAGETWO-80591 to track progress on the issue.

**Comment by sylvainraye** [comment]
I'm working on it #SQUASHTOBERFEST

**Comment by diglin** [comment]
The problem happens with the full page cache active

**Comment by okorshenko** [comment]
The issue has been fixed and delivered to 2.2-develop branch. Will be available with 2.2.2 release

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-81995

**Comment by magento-engcom-team** [comment]
Hi @koenner01. Thank you for your report. 
The issue has been fixed in magento-engcom/magento2ce#1289 by @magento-engcom-team in 2.3-develop branch
Related commit(s):
 - b41594122f532f6a89165a70906a5b257562c0df
 - 7369261d82e0e5ad926eaf4ccb2a8def79c4fcbb
 - dce1c66301cdddfee2dc2887f584c333d0b438b2
 - 95ed67ee4adbbd7f3f0107a52da4b9238cf44803
 - 05fcdd1a52c674ea09a34960cea71e73488fbc99
 - 8b1a2972b3f0ddd4713d00055b8d60f737a72ce4
 - 90addb731c3b253d3da4dc564ddf94ac3525bd0f
 - d5c9eef075cd5f4d9781ec7f3151a14e087d061d
 - 6ec62717c16ed96d45e670d7f867b724f0449aaa
 - f0af87edad79b295b4be87f6f92502bacfbc2764
 - 2497062749bf41c80aaebb9b6cd21eb48656f58c
 - 1f7347f47cca48238f9deabd48a0a280bc0c309f
 - 0d151c41539ba7c4d20ede20d614bb4be257903c
 - 07982fdea19074788654c6e825874cb8188ab627
 - 617da6ca08607890779ce71530a2871438c88351
 - 39ed21f83071ebffd71963e9097ecb59fd906702
 - 9aa036dec42211867f68f714a746b71395c43bb2
 - fda37ef0af75e6b28d4a0da1781b19c7a7894a97
 - 47b109180b6b5e499ba8d82e6c67888fca21ffe8
 - ef6badace5c6b046047a6e5f617020f9d4410a82
 - 809ec050986fa3ddfeec4a715ce5ace1e41a4427

The fix will be available with the upcoming patch release.

**Comment by sidolov** [comment]
Hi @koenner01. Thank you for your report. 
The issue has been fixed in magento/magento2#17243 by @mageprince in 2.1-develop branch
Related commit(s):
 - [c5d85e0cfbd7e38461a782c43eef69805a430b76](https://github.com/magento/magento2/commit/c5d85e0cfbd7e38461a782c43eef69805a430b76)
 - [f9696c964d21a3cfef60cffa6b23edf9a1a6be59](https://github.com/magento/magento2/commit/f9696c964d21a3cfef60cffa6b23edf9a1a6be59)

The fix will be available with the upcoming 2.1.15 release.
